lib/switchtec.c

```diff
@@ -652,33 +673,215 @@ int switchtec_hard_reset(struct switchtec_dev *dev)
+			if (num_entries >= LOG_ENTRIES_MAX) {
+				fprintf(stderr, "Invalid number of log entries: %d\n", num_entries);
+				return -1;
+			}
+			log_defs[mod_id].num_entries = num_entries;
+
+			strncpy(log_defs[mod_id].mod_name, mod_name, FW_MOD_NAME_LEN_MAX);
+
+			/* extract each log entry for the module */
+			for (entry_idx = 0; entry_idx < num_entries; entry_idx++) {
+				if (fgets(line, sizeof(line), log_def_file) == NULL) {
+					fprintf(stderr, "Failed to get log entry %d for module %s\n", entry_idx, log_defs[mod_id].mod_name);
+					return -1;
+				}
+
+				snprintf(log_defs[mod_id].entries[entry_idx], LOG_ENTRY_LEN_MAX, "%s", line);
+			}
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Parse an app log and write the results to a file
+ * @param[in] log_fd         - log output file descriptor
+ * @param[in] log_data       - logging data
+ * @param[in] count          - number of entries
+ * @param[in] init_entry_idx - index of the initial entry
+ * @param[in] log_defs       - log definitions
+ * @return 0 on success, negative value on failure
+ */
+static int write_parsed_log(int log_fd, struct log_a_data log_data[], size_t count,
+							int init_entry_idx, struct module_log_defs *log_defs)
+{
+	FILE *log_file;
+    int i;
+	int entry_idx = init_entry_idx;
+    unsigned long long time;
+    unsigned int nanos, micros, millis, secs, mins, hours, days;
+    unsigned int entry_num;
+    unsigned int mod_id;
+    unsigned int log_sev;
+	char *log_sev_str;
+
+	log_file = fdopen(log_fd, "w");
+	if (log_file == NULL) {
+		fprintf(stderr, "Unable to open output file\n");
+		return -1;
+	}
+
+	if (entry_idx == 0) {
+		fputs("   #|Timestamp                |Module       |Severity |Event\n", log_file);
+	}
+
+    for (i = 0; i < count; i ++) {
+        /* timestamp is in the first 2 DWords */
+        time = (((unsigned long long)log_data[i].data[0] << 32) | log_data[i].data[1]) * 10ULL;
+        nanos = time % 1000;
+        time /= 1000;
+        micros = time % 1000;
+        time /= 1000;
+        millis = time % 1000;
+        time /= 1000;
+        secs = time % 60;
+        time /= 60;
+        mins = time % 60;
+        time /= 60;
+        hours = time % 24;
+        days = time / 24;
+
+        /* log entry number, module ID, and log severity are in the 3rd DWord */
+        entry_num = log_data[i].data[2] & 0x0000FFFF;
+        mod_id = (log_data[i].data[2] >> 16) & 0xFFF;
+        log_sev = (log_data[i].data[2] >> 28) & 0xF;
+
+		/* validate the module ID */
+		if ((mod_id > FW_MOD_ID_MAX) ||
+		    (strlen(log_defs[mod_id].mod_name) == 0)) {
+			fprintf(log_file, "(Invalid module ID: %d)\n", mod_id);
+			continue;
+		}
+
+		/* validate the entry number */
+		if ((entry_num >= log_defs[mod_id].num_entries) ||
+		    (strlen(log_defs[mod_id].entries[entry_num]) == 0)) {
+			fprintf(log_file, "(Invalid log entry number: %d)\n", entry_num);
+			continue;
+		}
+
+		/* stringify and validate the log severity  */
+        switch (log_sev) {
```

Comment:
Maybe it's easier if we use something like this?

```
const char *sev_strs= {"DISABLED", "HIGHEST", ..."LOWEST"};
...
if(log_sev > 5) {
    fprintf(log_file, "(Invalid log severity: %d)\n", log_sev);
    continue;
}
log_sev_str = sev_strs[log_sev];
```

Reply:
Sure, I'll do that.